File: Companies/Google/EvaluateReversePolishNotation.java
package Companies.Google;

import java.util.Stack;

public class EvaluateReversePolishNotation {
    public int evalRPN(String[] tokens) {
        Stack<Integer> s = new Stack<>();
        for (String t : tokens) {
            if (t.equals("+")) {
                s.push(s.pop() + s.pop());
            } else if (t.equals("-")) {
                int l1 = s.pop();
                int l2 = s.pop();
                s.push(l2-l1);
            } else if (t.equals("*")) {
                s.push(s.pop() * s.pop());
            } else if (t.equals("/")) {
                int l1 = s.pop();
                int l2 = s.pop();
                s.push(l2/l1);
            } else {
                s.push(Integer.valueOf(t));
            }
        }
        return s.pop();
    }

}

File: Topics/Design/al/Evaluate_Reverse_Polish_Notation.cpp
using namespace std;

// A vector<int> can act as a stack and may be simpler:  right = values.back(); values.pop_back();
class EvaluateReversePolishNotation {
public:
    int evalRPN(const vector<string>& tokens) {
        const unordered_map<string, function<int(int, int)>> operators = {
            {"+", [](int a, int b) { return a + b; }},
            {"-", [](int a, int b) { return a - b; }},
            {"*", [](int a, int b) { return a * b; }},
            {"/", [](int a, int b) { return a / b; }}
        };

        stack<int> values;

        for (const string& token : tokens) {
            auto it = operators.find(token);
            if (it == operators.end()) {
                values.push(stoi(token));
                continue;
            }

            int right = values.top();
            values.pop();

            int left = values.top();
            values.pop();

            values.push(it->second(left, right));
        }

        return values.top();
    }
};
